### PR TITLE
lxd: Fix handling of images in projects with features.images=false

### PR DESCRIPTION
**Pull request by Fenixoff**
All projects with `features.images=false` can use all images added to the "default" project.
If these projects also have `features.profiles=true`, their default profiles should be associated
with all created images, otherwise
```
Error: Failed instance creation: Failed creating
instance record: Failed initialising instance: Failed getting root disk: No root device could be found
```
will arise on attempt to create an instance from this image inside all projects (with `features.images=false` and
`features.profiles=true`) except "default" and the project the image was created from.

The same error will take place when `features.images` option changes from `true` to `false`.

Both of these cases was fixed.

### CODE REVIEW

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
The static analysis is failing but I'm not sure why.

**Comment by tomponline** [comment]
Please can you add a test for this into tests/project.sh? 

Thanks!

File: lxd/db/cluster/projects.go [hunk 1]
```diff
@@ -199,7 +199,7 @@ DELETE FROM projects_config WHERE projects_config.project_id = ?
 	return nil
 }
 
-// InitProjectWithoutImages updates populates the images_profiles table with
+// InitProjectWithoutImages populates the images_profiles table with
```
Comment (tomponline) [inline lxd/db/cluster/projects.go:202]: This change should be a separate commit really.

File: lxd/api_project.go [hunk 2]
```diff
@@ -707,13 +707,22 @@ func projectChange(d *Daemon, project *api.Project, req api.ProjectPut) response
...
-				err = cluster.DeleteProfile(context.TODO(), tx.Tx(), project.Name, projecthelpers.Default)
+				err = cluster.DeleteProfile(ctx, tx.Tx(), project.Name, projecthelpers.Default)
 				if err != nil {
 					return fmt.Errorf("Delete project default profile: %w", err)
 				}
 			}
 		}
 
+		if shared.StringInSlice("features.images", configChanged) {
+			if shared.IsFalse(req.Config["features.images"]) && shared.IsTrue(req.Config["features.profiles"]) {
```
[the remark this answers is not in the document]
Comment (Fenixoff) [reply lxd/api_project.go:718]: The reasons were consistency with  the code directly above the one I added and readability. Also it allows easily add some handling on reverse change of this option. But nothing strict, I can rewrite it to one “if” statement if you wish.

[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/api_project.go:718]: Yes adding reverse change logic makes sense.

**Review by tomponline — CHANGES_REQUESTED** [review]
Overall this looks great, thank you!

There's a couple of nits, and a test would be good to catch any future regressions.

**Comment by tomponline** [comment]
If you rebase on master that should fix the static analysis failures.

File: lxd/db/cluster/projects.go (hunk 1 above)
Comment (Fenixoff) [reply lxd/db/cluster/projects.go:202]: Got it, I'll split it.

File: lxd/api_project.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fenixoff) [reply lxd/api_project.go:718]: So, can we leave as it is? Or should I convert to 1 `if` statement?

[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/api_project.go:718]: Let's move it to 1 line for now thanks

**Comment by tomponline** [comment]
jenkins: test this please

File: lxd/api_project.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fenixoff) [reply lxd/api_project.go:718]: Ok

File: test/suites/projects.sh [hunk 3]
```diff
@@ -429,9 +429,16 @@ test_projects_images_default() {
...
-  # The project can see images from the defaut project
+  # Create another project, without the features.images config.
+  lxc project create bar
+  lxc project set bar "features.images" "false"
+
+  # The project can see images from the default project
   lxc image list | grep -q testimage
 
+  # The image from the default project has correct profile assigned
+  ! lxc image show testimage | grep -qF "profiles: []" || false
```
Comment (tomponline) [inline test/suites/projects.sh:440]: I think we should test explicitly for the presence of the default profile using something like:

```
fingerprint="$(lxc image list -c f --format json | jq -r .[0].fingerprint)"
lxc query /1.0/images/"${fingerprint}"?project=test1 | jq -r .profiles[0] | grep -Fx default 
```

**Review by tomponline — CHANGES_REQUESTED** [review]
LGTM, lets just tweak the tests to check explicitly for the right profile.

File: test/suites/projects.sh (hunk 3 above)
Comment (Fenixoff) [reply test/suites/projects.sh]: Done. Seems `-c f` has no effect together `--format json` by the way.

**Review by tomponline — APPROVED** [review]
LGTM thanks

**Comment by Fenixoff** [comment]
Sorry, had to push one more time to make static analyser happy (removed trailing spaces).

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
Thanks!
